More robust installation instrucitons for Ubuntu and Debian

Lockszmith-GH opened the pull request:
Following up on #69, I present here the results of my own personal experience with installing apt-manage.

I've manually tested the quick-install script using Docker hub's images for ubuntu:18.04, ubuntu:20.04, ubuntu:22.04 and ubuntu:latest, as well as debian:10, debian:11 and debian:latest.

Let me know if this PR is not acceptable or if it breaks any internal rules.

Lockszmith-GH committed:
Added details install steps to README.md

Lockszmith-GH committed:
Created a quick-install.sh script

jacobgkau commented on README.rst:
```diff
@@ -75,9 +75,46 @@ From System Package Manager
+
     sudo apt install python3-repolib
 
+If your distro does not yet supply the package, you can install the
```
I think the new section should have its own title, maybe `From Pre-Compiled .deb Package`. If you want to change the above `From System Package Manager` to `From System Repository` to make the difference more clear, you could do that.

Lockszmith-GH replied:
Agreed, done.

jacobgkau commented on README.rst:
```diff
@@ -75,9 +75,46 @@ From System Package Manager
+    PREREQ_PKG=(curl python3-gnupg python3-debian ca-certificates)
+    . <(grep -E '^(VERSION_)?ID=' /etc/os-release)
+
+    # Make sure we're using the right mix, debian requires some missing
```
`debian` should be capitalized in a comment.

I would split this comment's line after the comma so the next line isn't just one word. Also, change the comma to a semicolon since these are both independent clauses.

You used "you" above and "we" here; it might be good to make that consistent.

Lockszmith-GH replied:
Fixed the captialization and line length. As for You vs We, documentation has You, code comments has We. If this still bugs you, I'll modify this.

jacobgkau replied:
> As for You vs We, documentation has You, code comments has We.

That's fine.

jacobgkau commented on README.rst:
```diff
@@ -75,9 +75,46 @@ From System Package Manager
 
+If your distro does not yet supply the package, you can install the
+[latest release](https://github.com/pop-os/repolib/releases/) using the
+steps below (At the time of writing this, the version is 2.0.0):
```
Change "steps below" to "script below" since you're using `apt-get` instead of `apt` as if this is a machine script rather than individual commands for a user to run.

jacobgkau commented on README.rst:
```diff
@@ -75,9 +75,46 @@ From System Package Manager
+
+    curl -Lo "/tmp/$PKG" "$URL"
+
+    # If we're on debian, zstd compression isn't supported by pkg.
```
`debian` should again be capitalized.

jacobgkau commented on README.rst:
```diff
@@ -75,9 +75,46 @@ From System Package Manager
+    curl -Lo "/tmp/$PKG" "$URL"
+
+    # If we're on debian, zstd compression isn't supported by pkg.
+    # So we repackage.
```
This is a fragment. Change the above period to a comma and un-capitalize this line.

Lockszmith-GH replied:
Changed the phrasing a bit, should be better/cleare now.

jacobgkau commented on README.rst:
```diff
@@ -75,9 +75,46 @@ From System Package Manager
+    # If we're on debian, zstd compression isn't supported by pkg.
+    # So we repackage.
+    if [[ "$ID" == 'debian' ]]; then
+        echo "We're on debian, so repackaging without zstd compression..."
```
Maybe change `We're on debian` to `Running on Debian`.

jacobgkau reviewed (changes requested):
Thanks for the contribution! The formatting currently looks broken:

![image](https://github.com/pop-os/repolib/assets/7199422/a7a20bc4-3239-4152-9d72-d817791b9155)

I think you want to put this in a code block rather than a quote. Instead of using four spaces, try putting the block in between two lines of three backticks (```).

I've made some style notes. The script looks reasonable to me, so I'll test once the formatting is fixed.

Lockszmith-GH committed:
Updateding README.rst (for PR #70)

Addressing [notes](https://github.com/pop-os/repolib/pull/70#pullrequestreview-1568207274) from @jacobgkau about the formatting and content.

Lockszmith-GH commented:
I wasn't familiar with RST enough when wrote it, and was focused on dumping my mind first.
Now it should be better.

Lockszmith-GH committed:
Added the comments from the documentation to the quick-install script.

jacobgkau commented on README.rst:
```diff
@@ -71,13 +71,59 @@ Allows changing configuration details of a given repository
+------------------------------
+
+If your distro does not yet supply the package, you can install the
+`latest release`__ using the script below (At the time of writing this,
```
The parenthetical not at the beginning of a sentence needs to have its first word un-capitalized. The current wording is a little overpersonal anyway, so please replace it with this:

> (replace `2.0.0` with the version you wish to install):

Lockszmith-GH replied:
Shouold be better now.

jacobgkau reviewed (changes requested):
Looks much better. Found one more small formatting tweak (I don't have permission to push a commit to your fork to do it myself.)

Lockszmith-GH committed:
Improve phrasing around package version

jacobgkau commented on README.rst:
```diff
@@ -71,13 +71,60 @@ Allows changing configuration details of a given repository
+
+If your distro does not yet supply the package, you can install the
+`latest release`__ using the script below
+(replace ``PKG_VER='2.0.0'`` with the version you wish to install):
```
This isn't accurate. You don't replace `PKG_VER='2.0.0'`, you replace only the version number of `2.0.0` because the variable still needs to get created. Please update to the wording I previously suggested:

> (replace `2.0.0` with the version you wish to install):

jacobgkau commented on quick-install.sh:
```diff
@@ -0,0 +1,65 @@
+#! /usr/bin/env bash
```
I just noticed this file. Why do we need the same script in both the README and a separate file? This seems like a packaging thing more than something that needs to be in this repository. Having documentation doesn't hurt, but having a remotely-runnable script for other OSes seems a little out of scope.

Conversely, if the file is here, why are we duplicating the script in the README instead of just linking to the file?

Lockszmith-GH replied:
In my initial question in #69 I suggested to improve the documentation and optionally provide a script.
I posted both, as I wasn't clear on what path would be preferred.

I'm more than OK to change the documentation to point to the convenience script instead.

jacobgkau replied:
I see. It sounds like Levi was saying instructions in the README are welcome but that the existing PyPI installation method is intended to fill the role of a convenient installation script. I don't think this is really ruffling any feathers, though. It makes sense to have a complete script be in its own file so it's not necessary to copy/paste part of the README out into a different file before running it.

If you want the script to be in its own file, then update the README section's wording to this:

-----

From Pre-Compiled .deb Package
------------------------------

If your distro does not yet supply the package, you can download and install the .deb package from the [latest release](https://github.com/pop-os/repolib/releases/latest). A script to do this automatically can be found in [quick-install.sh]().

jacobgkau commented on quick-install.sh:
```diff
@@ -0,0 +1,65 @@
+    apt-get update
+
+    # Try differnt ways of installing apt-manage
+    # First let's hope there is package ready for installation
```
Are you aware of any Debian-based distributions that currently package repolib aside from Pop!_OS (where it's installed by default)?

A script shouldn't "hope" anything. If this file is kept, please replace the above two comment lines with:

> \# First, try installing from the system repositories.

Lockszmith-GH replied:
I was under the impression that Pop!_OS submit such packages to upstream distros.
I'll drop the test.

jacobgkau replied:
The test makes sense logically, it just seems superfluous at this time. If the script stays, I think it's fine for the test to be there, with the modified comment.

> I was under the impression that Pop!_OS submit such packages to upstream distros.

The main things we would generally submit ourselves to upstream projects would be patches to enable hardware support for computers that we sell/support, and bug fixes or other contributions to projects that already exist upstream. A lot of Pop!_OS tooling is going to be Pop!_OS-specific and not very interesting to e.g. Ubuntu or Debian. Repolib has support for `popdev:` sources to add staging repositories generated from the Pop!_OS GitHub organization, which aren't always going to work well on other distros, for example. (We'd probably accept contributions to conditionally enable features like that on Pop!_OS if someone was interested in packaging for other distros, but we haven't had the need to do that ourselves.)

Lockszmith-GH replied:
I'll leave the test, only for the possibility to use within Dockerfiles that might use Pop!_OS or other Debian based distros.

jacobgkau commented on quick-install.sh:
```diff
@@ -0,0 +1,65 @@
+    # First let's hope there is package ready for installation
+    apt-get install --yes apt-manage || true
+
+    # If it still doesn't exist, let's get it from repo
```
The term "repo" is ambiguous and also applies to the distribution package repo that would be used above. I would also add punctuation since most of the other comments have it.

> \# If it still doesn't exist, download the package from GitHub.

Lockszmith-GH replied:
makes total sense, fixing this now.

Lockszmith-GH committed:
Moving focus to convenience script

jacobgkau commented on README.rst:
```diff
@@ -71,13 +71,26 @@ Allows changing configuration details of a given repository
+.. code:: shell
+
+    sudo apt update && sudo apt install --yes --no-install-recommends curl
+    sudo bash <(curl -s https://raw.githubusercontent.com/pop-os/repolib/HEAD/quick-install.sh)
```
I don't want our README telling people to run a web script as root. Please update the wording for this section to what I suggested in https://github.com/pop-os/repolib/pull/70#discussion_r1294916406.

jacobgkau commented on quick-install.sh:
```diff
@@ -0,0 +1,76 @@
+    fi
+fi
+
+# Passing '-' as the sole argument, will not run apt-manage, otherwise
```
The first comma definitely shouldn't be there. "the any argument" in the next line also looks like a typo.

I'm confused about what this three-line comment is saying in general.

Lockszmith-GH replied:
I'll remove that last part, and replace with a 'success/fail' message, this has no place on a public repo, as it is more to my personal use cases. Sorry about that (I should go with kiss)

jacobgkau commented on quick-install.sh:
```diff
@@ -0,0 +1,76 @@
+
+        apt-get update
+        # For the odd case where /tmp is missing
+        # (might happen in some obscure docker images)
```
Please capitalize Docker in the comment.

Lockszmith-GH committed:
Cleanup per notes from PR

Lockszmith-GH commented:
BTW: The way I've tested the script:

Dockerfile:

```Dockerfile
ARG  FROM_TAG=debian:latest
FROM ${FROM_TAG}

RUN apt-get update \
    && apt-get install --yes --no-install-recommends curl ca-certificates

COPY ./quick-install.sh .
RUN bash ./quick-install.sh

ENTRYPOINT ["/usr/bin/apt-manage"]
CMD ["--help"]
```

script, that iterates on different image tags, and builds the Dockerfile:

```bash
#! /usr/bin/env bash

set -e

YY=$(date +%y)
TAGS=(${@:- \
    debian:latest debian:12-slim debian:11-slim debian:10-slim\
    debian:oldstable-slim debian:sid-slim \
    ubuntu:latest ubuntu:${YY}.04 ubuntu:$(( YY - 1 )).04 ubuntu:devel
})

cp ~/_code/repolib_fork4pr/quick-install.sh .

# Prefer to pull, so that clearing buildx cache doesn't get rid of images.
xargs -rn1 docker pull <<<"${TAGS[*]}"

for FROM_TAG in "${TAGS[@]}"; do
    TAG="${FROM_TAG//:/-}"
    printf "Building %s...\n" "test-repolib-install:${TAG}"
    docker build . --build-arg "FROM_TAG=$FROM_TAG" --tag "test-repolib-install:${TAG}"
    docker run --rm "test-repolib-install:${TAG}"
    printf "\nBuilding %s complete.\n\n" "test-repolib-install:${TAG}"
done
```

jacobgkau commented on README.rst:
```diff
@@ -71,13 +71,22 @@ Allows changing configuration details of a given repository
+
+If your distro does not yet supply the package, you can download and
+install the .deb package from the `latest release`__.  
+A script to do this automatically can be found in `quick_install.sh`__ script.
```
You added a word, and now it's grammatically incorrect. To be grammatically correct, it either needs to be "found in [quick-install.sh]()." as I [asked](https://github.com/pop-os/repolib/pull/70#discussion_r1294916406), or "found in the [quick-install.sh]() script." I would prefer the version I asked for since we already use the word "script" earlier in the sentence.

Lockszmith-GH replied:
sorry about that - I have the oldscool habbit of dropping the 'the' - that's what I get from learning from oldschool british style english teachers back in the day.

Lockszmith-GH committed:
grammar correction

jacobgkau reviewed (approved):
Confirmed the script works on Debian 12.1 and Ubuntu 22.04 LTS. (`curl` is not present by default in either OS, so I had to install that before using the suggested remote-run command.)

Mentioning that pre-compiled .debs are available before suggesting compiling it manually makes sense, separately from the script.

Lockszmith-GH commented:
> ...`curl` is not present by default in either OS, so I had to install that before using the suggested remote-run command.

Your comment is right, but the README.md no longer supplies the curl cmd, and the suggestion is in the script file itself - which means it's for people who can handle it.


> Mentioning that pre-compiled .debs are available before suggesting compiling it manually makes sense, separately from the script.

I'll add a blurb around this.

Lockszmith-GH committed:
Improve comment

Lockszmith-GH committed:
Improve comment

Lockszmith-GH committed:
Add a note about installing deb manually

Lockszmith-GH commented:
Let me know if you feel anything else should be added.

jacobgkau commented:
No further changes were needed. I was stating something that was already present in support of my approval.

jacobgkau reviewed (changes requested):
Please change back to the version that was previously approved. We don't need to explain how to install a `.deb`, we have a GUI for that and there are multiple different commands people might want to use anyway.

Lockszmith-GH commented:
> Please change back to the version that was previously approved. We don't need to explain how to install a `.deb`, we have a GUI for that and there are multiple different commands people might want to use anyway.

Sorry about that - I'll revert those commits.

Lockszmith-GH committed:
Revert "Add a note about installing deb manually"

This reverts previous unnecessary commit.
